qkeras/codebook.py

```diff
@@ -0,0 +1,205 @@
+  km_models = [KMeans(2**bits)] * len(activation_indexes)
+  cb_tables = [[]] * len(activation_indexes)
+  models = []
+  x = x_in = model.layers[0].output
+  for i in range(1, len(model.layers)):
+    layer = model.layers[i]
+    x = layer(x)
+    if i in activation_indexes or i == len(model.layers) - 1:
+      print("\nCreating submodel...")
+      models.append(Model([x_in], [x]))
+      x = x_in = Input(layer.output[0].shape,
+                       batch_size=layer.output.shape[0],
+                       dtype=layer.output.dtype)
+      models[-1].compile(**compile_config)
+      print(models[-1].summary())
+  print('\nsample_size: ', sample_size)
+  x = X_train
+  for i, model in enumerate(models[:-1]):
+    print(f'fitting km[{i}]...')
+    x = model.predict(x)
+    km = km_models[i]
+    temp = x.flatten().reshape(-1, 1)
+    if sample_size < 1.0:
+      idxs = np.random.choice(x.shape[0], size=int(sample_size * x.shape[0]))
+      temp = temp[idxs]
+    km.fit(temp)
+    quantizer = getattr(model.layers[-1], 'quantizer',
+                    getattr(model.layers[-1], 'activation'))
+    km.cluster_centers_ = quantizer(km.cluster_centers_).numpy()
+    km.cluster_centers_.sort(axis=0)
+    cb_tables[i] = create_in_out_table(km, quantizer)
+  x = X_test
+  for i, model in enumerate(models[:-1]):
+    x = model.predict(x)
+    km = km_models[i]
+    x = km.cluster_centers_[km.predict(x.flatten().reshape(-1, 1))].reshape(x.shape)
+    n_unique = np.unique(x.flatten()).shape[0]
+    print(f"Number of unique activations: {n_unique}")
+    assert n_unique <= 2**bits
+
+  print('\nEvaluating...')
+  models[-1].evaluate(x, y_test, verbose=2)
+  return cb_tables, models, km_models
+
+
+def weight_compression(weights, bits, axis=0, quantizer=None):
+  """ Creates an in, out table that maps weight values to their codebook values.
+  Based on the idea presented by https://arxiv.org/pdf/1911.02079.pdf
+
+  Arguments:
+    weights: Numpy array
+    bits: Number of bits to compress weights to. This will
+      results in 2**bits codebook values
+    axis: axis to apply quantization by
+    quantizer: quantizer function that will be applied to codebook values
+
+  Returns:
+    km_models: list of fitted KMeans algorithms
+    quantized_embeddings: Numpy array MxN
+  """
+  assert bits <= 8
+  n = 2**bits
+  index_table = []
+  codebook_table = np.zeros((weights.shape[axis], n))
+  km_models = [None] * weights.shape[axis]
+
+  for i, w in tqdm(enumerate(np.split(weights, weights.shape[axis], axis))):
+    original_shape = w.shape
+    w = w.ravel()
+    km = KMeans(n)
+    km.fit(w.reshape(-1, 1))
+    if quantizer:
+      km.cluster_centers_ = quantizer(km.cluster_centers_).numpy()
+    km.cluster_centers_.sort(axis=0)
+
+    km_models[i] = km
+    codebook_table[i, :] = km.cluster_centers_.flatten()
+    preds = km.predict(w.reshape(-1, 1))
+    index_table.append(preds.reshape(original_shape))
+  
+  index_table = np.concatenate(index_table, axis)
```

Comment:
regarding index_table, is the shape of index_table the same as the shape of weight?

For example, a conv layer, we should have weights[kh, kw, k_in, k_out] approximates or equals codebook_table[axis][index_table[kh, kw, k_in, k_out]]

In your jupyter notebook, could you verify that new_conv_weights closes to/equals conv_weights around your code
```
new_conv_weights = np.zeros(conv_weights.shape)
for i in range(conv_weights.shape[axis]):
  new_conv_weights[:,:,:,i] = codebook_table[i][index_table[:,:,:,i]]
```

Reply:
1. yes, index_table is the same shape of weights.
Here are the stats from the absolute difference from both weight matrices
"sp.stats.describe((abs(new_conv_weights - conv_weights)).ravel())"
DescribeResult(nobs=864, minmax=(1.3589859008789062e-05, 0.33330094814300537), mean=0.05314029095818653, variance=0.0026912063707205084, skewness=2.260220215254187, kurtosis=6.558495502282023)
2. After codebook quantization from 4 to 3 bits: val_acc: 0.6202 . Original quantized model val_acc: 0.6471.

Reply:
Hi Jose, if possible, could you write a codebook_test.py to assert those numbers? Thanks.

Reply:
Yeah I can add it tomorrow or Wednesday the latest.